sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -927,6 +942,185 @@ abstract class GpuBaseAggregateMeta[INPUT <: SparkPlan](
+   * At last, we traverse aggregateAttributes and resultExpressions, overriding data type in
+   * RapidsMeta if necessary, in order to ensure TypeChecks tagging exact data types in runtime.
+   */
+  private def overrideAggBufTypes(): Unit = {
+    val desiredAggBufTypes = mutable.HashMap.empty[ExprId, DataType]
+    val desiredInputAggBufTypes = mutable.HashMap.empty[ExprId, DataType]
+    // Collects exprId from TypedImperativeAggBufferAttributes, and maps them to the data type
+    // of `TypedImperativeAggExprMeta.aggBufferAttribute`.
+    agg.aggregateExpressions.zipWithIndex.foreach {
+      case (expr, i) if expr.aggregateFunction.isInstanceOf[TypedImperativeAggregate[_]] =>
+
+        val aggFn = expr.aggregateFunction
+        val aggMeta = aggregateExpressions(i).childExprs.head
+            .asInstanceOf[TypedImperativeAggExprMeta[_]]
+        val desiredDataType = aggMeta.aggBufferAttribute.dataType
+
+        var buf = aggFn.aggBufferAttributes.head
+        desiredAggBufTypes(buf.exprId) = desiredDataType
+
+        buf = aggFn.inputAggBufferAttributes.head
+        desiredInputAggBufTypes(buf.exprId) = desiredDataType
+
+      case _ =>
+    }
+
+    // Overrides the data types of typed imperative aggregation buffers for type checking
+    aggregateAttributes.foreach { attrMeta =>
+      attrMeta.wrapped match {
+        case ar: AttributeReference if desiredAggBufTypes.contains(ar.exprId) =>
+          attrMeta.overrideDataType(desiredAggBufTypes(ar.exprId))
+        case _ =>
+      }
+    }
+    resultExpressions.foreach { retMeta =>
+      retMeta.wrapped match {
+        case ar: AttributeReference if desiredInputAggBufTypes.contains(ar.exprId) =>
+          retMeta.overrideDataType(desiredInputAggBufTypes(ar.exprId))
+        case _ =>
+      }
+    }
+  }
+}
+
+object GpuTypedImperativeSupportedAggregateExecMeta {
+
+  private val entireAggFallbackCheck = TreeNodeTag[Boolean](
+    "rapids.gpu.checkAndFallbackAggregateExecEntirely")
+
+  private def checkAndFallbackEntirely(
+      meta: GpuTypedImperativeSupportedAggregateExecMeta[_]): Unit = {
+    // We only run the check for final stages which contain TypedImperativeAggregate.
+    val needToCheck = meta.agg.aggregateExpressions.exists(e =>
+      (e.mode == Final || e.mode == Complete) &&
+          e.aggregateFunction.isInstanceOf[TypedImperativeAggregate[_]])
+    if (!needToCheck) return
+    // Avoid duplicated check and fallback.
+    val checked = meta.agg.getTagValue[Boolean](entireAggFallbackCheck).contains(true)
+    if (checked) return
+
+    meta.agg.setTagValue(entireAggFallbackCheck, true)
+    val logicalPlan = meta.agg.logicalLink.get
+    val stageMetas = mutable.ListBuffer[GpuBaseAggregateMeta[_]]()
+    // Go through all Aggregate stages to check whether all stages is GPU supported. If not,
+    // we fall back all GPU supported stages to CPU.
+    if (recursiveCheckForFallback(meta, logicalPlan, stageMetas)) {
```

Comment:
To adapt AQE, I take advantage of  `gpuSupportedTag` which was introduced by @andygrove. I added the line
```scala
wrapped.getTagValue(gpuSupportedTag).foreach(_.foreach(willNotWorkOnGpu))
```
in `GpuTypedImperativeSupportedAggregateExecMeta.tagPlanForGpu` to retrieve the information about GPU support which is captured and cached during `GpuQueryStagePrepOverrides`.

Reply:
I understand, but does that run over the entire plan at some point, or is it just sections of the plan.  If it is the entire plan it would be good to explain that in a comment, because otherwise it looks like we have cases where we can crash.

Reply:
I made up some comments on this section.